contributors/design-proposals/flakiness-sla.md

```diff
@@ -0,0 +1,92 @@
+# Kubernetes Testing Flakiness SLA
+
+This document captures the expectations of the community about flakiness in
+our tests and our test infrastructure. It sets out an SLA (Service Level 
+Agreement) for flakiness in our tests, as well as actions that we will
+take when we are out of SLA.
+
+
+## Definition of a "Flake"
+
+We'll start by the definition of a _flake_. _Flakiness_ is defined for a
+complete run of one required job in the pre-submit testing infrastructure 
+(e.g. pull-kubernetes-e2e-gke-gci) for a given pull request (PR). We will
+not measure flakiness SLA for e2e jobs that are not required for code merge.
+
+A pre-submit job's test result is considered to
+be a flake according to two criteria:
+
+1) it both fails and passes without any changes in the code
+or environment being tested.
+
+2) the PR in question doesn't cause the flake itself.
+
+### Measuring flakiness
+There are a number of challenges in monitoring flakiness. We expect that
+the metric will be heuristic in nature and we will iterate on it over time.
+Identifying all of the potential problems and ways to measure the metric
+are out of the scope of the document, but some currently known challenges
+are listed at the end of the document.
+
+## Flakiness SLA
+We will measure flakiness based on pull requests that are run through pre-submit
+PR testing. The metric that we will monitor is: Pre-submit flakes per PR. This metric
+will be calculated on a daily and weekly basis using:
+
+Sum(Flakes not caused by the PR) / Total(PRs)
+
+Our current SLA is that this metric will be less than 0.01
+(1% of all PRs have flakes that are not caused by the PR itself).
+
+## Activities on SLA violation
+When the project is out of SLA for flakiness on either the daily or weekly metric
+we will stop the merge queue.
```

Comment:
tl;dr; +1 to @fejta 

Two issues remain unresolved in this proposal:

1. Metric & SLA - We continue to iterate on the metric and the goal, I think we can go ahead and merge this with the expectation we will continue to do that until we are satisfied. We need to invest in tools to continuously measure, report, and act/alert on SLI. Target can gradually change.

2. Effectiveness of closing the submit queue. 
Blocking the submit queue for everyone would not solve the problem of building ownership of the problematic tests, and would have an excessively negative impact on community productivity and morale.
If we remove this from the proposal I am ok with merging it. 
I have drafted an alternative suggestion, just as an example.
It should be also discussed in the SIG like @fejta suggested.

- Assign test suites/flakes to owners/SIGs. I think it's important that the SIGs accept responsibility/ownership, unlike the previous attempt at random assignment.
- Monitor flakes, initial response and resolution time %  by SIG and build a dashboard to display the data over time.
- Define SLO on initial response to flake (e.g. 3 business days), with the goal of meeting the desired timeframe for 80% of the instances. 
- Define SLO to resolution (e.g. 2 weeks), with the goal of meeting the desired timeframe for 60% of the instances. 
- Present status each community meeting by SIG and have SIG report back. Previously we have done it in the burndown meetings and it worked. Doing it regularly I expect will have positive impact again.
- Before transitioning SLOs to SLAs, we need to draft expectations from SIGs. Each SIG needs to submit their plan how they would meet SLOs.
- Transition SLOs to SLAs with an agreed consequences, by SIG. e.g. blocking merging of new PRs from that SIG 

I acknowledge this is also not perfect. This imply changing the character of some of the SIGs from just discussions to more tactical work and accountability, but I think this is progress in the right direction empowering SIGs. This also requires building tools to support it.